### PR TITLE
Rk/improvement/core data update

### PR DESCRIPTION
**Pull request by kapaljaka**
Mostly code-clean up. Didn't change core data stack almost at all. Should be done in other PR, because there are too much changes already and it's better to review and merge now.

### CODE REVIEW

File: PetrMazepa/Source/Article.swift [hunk 1]
```diff
@@ -38,4 +43,67 @@ class Article : NSObject {
...
+        guard
+            let aElement = element.searchWithXPathQuery("//a").first as? TFHppleElement,
+            let href = aElement.attributes["href"] as? String else {
+                return nil
+        }
+        
+        let imgElement = element.firstChildWithTagName("img")
+        
+        guard
+            let identifier = identifierFromHref(href),
```
Comment (artemstepanenko) [inline PetrMazepa/Source/Article.swift]: Wouldn't it better to move this condition into the previous guard?

File: PetrMazepa/Source/ContentProvider.swift [hunk 2]
```diff
@@ -100,11 +98,9 @@ class ContentProvider: ArticleStorage, FavouriteArticlesStorage, ArticlesFetcher
...
             
             if let notNilDetails = details {
-                
-                self.coreData.saveArticleDetails(notNilDetails, article: article)
-                self.coreData.saveContext()
-                article.saved = true
-                
+
+                self.coreData.context.performChanges({ self.coreData.saveArticleDetails(notNilDetails, article: article) })
+                article.saved = true
```
Comment (artemstepanenko) [inline PetrMazepa/Source/ContentProvider.swift]: Shouldn't this line be inside the changes closure?

File: PetrMazepa/Source/CoreDataManager.swift [hunk 3]
```diff
@@ -166,88 +114,97 @@ class CoreDataManager : FavouriteArticlesStorage, FavouriteMaker {
...
-        return Article(id: moArticle.id!, title: moArticle.title!, thumbPath: moArticle.thumbPath!, saved: moArticle.details != nil, favourite: moArticle.favourite!.boolValue, topOffset: moArticle.topOffset!.floatValue)
-    }
-    
-    private func detailsFromManagedObject(moDetails: MOArticleDetails) -> ArticleDetails {
-        return ArticleDetails(htmlText: moDetails.htmlText!)
+
+    func performChanges(block: () -> ()) {
+        performBlock { 
+            block()
+            self.saveOrRollback()
```
Comment (artemstepanenko) [inline PetrMazepa/Source/CoreDataManager.swift]: Isn't it better to save changes less often?

File: PetrMazepa/Source/ScreenFlow.swift [hunk 4]
```diff
@@ -52,7 +52,7 @@ class ScreenFlow: NSObject, ArticleDetailsPresenter, SettingsPresenter, SearchPr
     }
     
     func save() {
-        self.coreData.saveContext()
+        self.coreData.context.saveOrRollback()
     }
```
Comment (artemstepanenko) [inline PetrMazepa/Source/ScreenFlow.swift]: Since we update the persistent storage after every single change, this method is not needed anymore.

File: PetrMazepa/Source/Utils.swift [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+//
+
+import Foundation
+
+
+extension NSURL {
+    
+    static func temporaryURL() -> NSURL {
+        return try! NSFileManager.defaultManager().URLForDirectory(NSSearchPathDirectory.CachesDirectory, inDomain: .UserDomainMask, appropriateForURL: nil, create: true).URLByAppendingPathComponent(NSUUID().UUIDString)
+    }
```
Comment (artemstepanenko) [inline PetrMazepa/Source/Utils.swift]: To make it slightly more clear what this url really is, please, rename it to `cachesDirectoryURL()`

File: PetrMazepa/Source/Utils.swift [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    
+    static var documentsURL: NSURL {
+        return try! NSFileManager.defaultManager().URLForDirectory(.DocumentDirectory, inDomain: .UserDomainMask, appropriateForURL: nil, create: true)
+    }
```
Comment (artemstepanenko) [inline PetrMazepa/Source/Utils.swift]: `documentDirectoryURL()`?

File: PetrMazepa/Source/CoreDataManager.swift [hunk 7]
```diff
@@ -9,95 +9,50 @@
...
-    }()
-    
     lazy var context: NSManagedObjectContext = {
-
-        let coordinator = self.coordinator
-        var context = NSManagedObjectContext(concurrencyType: .MainQueueConcurrencyType)
-        context.persistentStoreCoordinator = coordinator
-        return context
+        return self.createMainContext()
     }()
```
Comment (artemstepanenko) [inline PetrMazepa/Source/CoreDataManager.swift]: I'm wondering, can you do this in one line?

File: PetrMazepa/Source/CoreDataManager.swift [hunk 8]
```diff
@@ -9,95 +9,50 @@
...
     
     func notFavoriteArticles() -> [Article] {
         return self.requestArticles(favorite: false)
     }
     
-    func favouriteArticles() -> [Article] {
-        return self.requestArticles(favorite: true)
-    }
-    
     func detailsFromArticles(article: Article) -> ArticleDetails? {
```
Comment (artemstepanenko) [inline PetrMazepa/Source/CoreDataManager.swift]: I noticed, the plural form of article doesn't make much sense here. I know, it's not your mistake, but couldn't you fix this?

File: PetrMazepa/Source/ManagedObjectType.swift [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+//
+//  Created by Roman Kopaliani on 6/9/16.
+//  Copyright © 2016 TramCoders. All rights reserved.
+//
+
+import Foundation
+import CoreData
+
+public class ManagedObject: NSManagedObject {
+}
```
Comment (artemstepanenko) [inline PetrMazepa/Source/ManagedObjectType.swift]: Why do we need the subclass?

**Comment by artemstepanenko** [comment]
In general: I don't like much that the context and managed objects are passed around the content provider. I would prefer them to be only beyond the interface of CoreDataManager.

File: PetrMazepa/Source/Article.swift (hunk 1 above)
Comment (kapaljaka) [inline PetrMazepa/Source/Article.swift]: Agreed. Updated.

File: PetrMazepa/Source/ContentProvider.swift (hunk 2 above)
Comment (kapaljaka) [inline PetrMazepa/Source/ContentProvider.swift]: Idk, didn't change much here, to not break anything.
That's actually the mane architecture-related (not talking about memory and performance now) problem why we cannot easily migrate to structs -> 'saved' property is set on Article class, not the backed NSMO, and it makes this class mutable.

File: PetrMazepa/Source/CoreDataManager.swift (hunk 3 above)
Comment (kapaljaka) [inline PetrMazepa/Source/CoreDataManager.swift]: yes it does, but why do you mean by that? I haven't changed places where calls for save are made, just changed save mech a little bit. 
Calls for saves are gonna be changed in separate PR where CoreData architecture is gonna be changed. It's just too many changes is this PR already, it's hard to review and hypothetically it might be hard to merge (not in this case though).

File: PetrMazepa/Source/ScreenFlow.swift (hunk 4 above)
Comment (kapaljaka) [inline PetrMazepa/Source/ScreenFlow.swift]: agreed, but I wanna to remove this kind of functions in separate PR if you don't mind.

File: PetrMazepa/Source/Utils.swift (hunk 5 above)
Comment (kapaljaka) [inline PetrMazepa/Source/Utils.swift]: Done

File: PetrMazepa/Source/Utils.swift (hunk 6 above)
Comment (kapaljaka) [inline PetrMazepa/Source/Utils.swift]: agreed, that's more explicit

File: PetrMazepa/Source/CoreDataManager.swift (hunk 7 above)
Comment (kapaljaka) [inline PetrMazepa/Source/CoreDataManager.swift]: nah, it looks ugly in one line
either way, it's not gonna last long, I'm planning to add another context and will remove lazy init for main context also (what's the point anyway of doing it, we need to from the start)

File: PetrMazepa/Source/CoreDataManager.swift (hunk 8 above)
Comment (kapaljaka) [inline PetrMazepa/Source/CoreDataManager.swift]: done

File: PetrMazepa/Source/ManagedObjectType.swift (hunk 9 above)
Comment (kapaljaka) [inline PetrMazepa/Source/ManagedObjectType.swift]: I am not sold on it myself tbh with you. Guys from objc.io says that it helps to separate concerns: we are gonna hide all the ManagedObjectType protocol extensions methods( there are already few of them and could be much in the future) for the all pure NSMO objects. But it might be an overkill. Should I remove it what do you think?

**Comment by kapaljaka** [comment]
Core Data Manager looks like a superclass right now, it has too much responsibilities – create stuck, insert objects, find, save (hypothetical delete, sync etc). 
What I propose we should do, if you really want to hide all the insides:
– create a Core Data stack object (set store, couple of contexts and all that jazz, but that's it)
– allows NSMO to fetch, insert etc. themselves in contexts with predicates (like it's done in this PR) 
– create an additional abstraction layer (call it Persistent layer if you want, VM doesn't need to know about exact persistence mechanism) which is gonna return simple classes (structs?), maybe implement a simple proxy like object filled with static methods, need to think about that more
– remove superclass CoreData manager, it's too big, gonna become bigger when added new entities (comments and all that shit), it's hard to maintain and it's hard to cover with tests.
